unpacker.go

```diff
@@ -87,85 +88,128 @@ func (u *unpacker) unpack(ctx context.Context, config ocispec.Descriptor, layers
+		})
+
+		var (
+			key    string
+			mounts []mount.Mount
+		)
+
+		for try := 1; try <= 3; try++ {
+			// Prepare snapshot with from parent, label as root
+			key = fmt.Sprintf("extract-%s %s", uniquePart(), chainID)
+			mounts, err = sn.Prepare(ctx, key, parent.String(), labelOpt)
+			if err != nil {
+				if errdefs.IsAlreadyExists(err) {
+					if _, err := sn.Stat(ctx, chainID); err != nil {
+						if !errdefs.IsNotFound(err) {
+							return errors.Wrapf(err, "failed to stat snapshot %s", chainID)
+						}
+						// Try again, this should be rare, log it
+						log.G(ctx).WithField("key", key).WithField("chainid", chainID).Debug("extraction snapshot already exists, chain id not found")
+					} else {
+						// no need to handle, snapshot now found with chain id
+						continue EachLayer
+					}
+				} else {
+					return errors.Wrapf(err, "failed to prepare extraction snapshot %q", key)
+				}
+			} else {
+				break
 			}
-			states[i].downloaded = true
-			break
 		}
-		for i := range states {
-			if !states[i].downloaded {
-				break
+		if err != nil {
+			return errors.Wrap(err, "unable to prepare extraction snapshot")
+		}
+
+		// Abort the snapshot if commit does not happen
+		abort := func() {
+			if err := sn.Remove(ctx, key); err != nil {
+				log.G(ctx).WithError(err).Errorf("failed to cleanup %q", key)
 			}
-			if states[i].unpacked {
-				continue
+		}
+
+		if fetchErr == nil {
+			fetchErr = make(chan error, 1)
+			fetchOffset = i
+			fetchC = make([]chan struct{}, len(layers)-fetchOffset)
+			for i := range fetchC {
+				fetchC[i] = make(chan struct{})
 			}
 
-			log.G(ctx).WithFields(logrus.Fields{
-				"desc": states[i].layer.Blob,
-				"diff": states[i].layer.Diff,
-			}).Debug("unpack layer")
+			go func() {
+				err := u.fetch(ctx, h, layers[i:], fetchC)
+				if err != nil {
+					fetchErr <- err
+				}
+				close(fetchErr)
+			}()
+		}
 
-			unpacked, err := rootfs.ApplyLayerWithOpts(ctx, states[i].layer, chain, sn, a,
-				u.config.SnapshotOpts, u.config.ApplyOpts)
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case err := <-fetchErr:
 			if err != nil {
 				return err
 			}
+		case <-fetchC[i-fetchOffset]:
+		}
 
-			if unpacked {
-				// Set the uncompressed label after the uncompressed
-				// digest has been verified through apply.
-				cinfo := content.Info{
-					Digest: states[i].layer.Blob.Digest,
-					Labels: map[string]string{
-						"containerd.io/uncompressed": states[i].layer.Diff.Digest.String(),
-					},
-				}
-				if _, err := cs.Update(ctx, cinfo, "labels.containerd.io/uncompressed"); err != nil {
-					return err
-				}
-			}
+		diff, err := a.Apply(ctx, desc, mounts)
+		if err != nil {
+			abort()
+			return errors.Wrapf(err, "failed to extract layer %s", diffIDs[i])
+		}
+		if diff.Digest != diffIDs[i] {
+			abort()
+			return errors.Errorf("wrong diff id calculated on extraction %q", diffIDs[i])
+		}
+
+		if err = sn.Commit(ctx, chainID, key, labelOpt); err != nil {
```

Comment:
concurrent download the same image and commit might return ErrAlreadyExist. for this case, we should skip it.